tachyon/math/finite_fields/square_root_algorithms/tonelli_shanks.h

```diff
@@ -0,0 +1,80 @@
+// Copyright 2022 arkworks contributors
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.arkworks and the LICENCE-APACHE.arkworks
+// file.
+
+#ifndef TACHYON_MATH_FINITE_FIELDS_SQUARE_ROOT_ALGORITHMS_TONELLI_SHANKS_H_
+#define TACHYON_MATH_FINITE_FIELDS_SQUARE_ROOT_ALGORITHMS_TONELLI_SHANKS_H_
+
+#include <utility>
+
+namespace tachyon::math {
+
+template <typename F>
+constexpr bool ComputeTonelliShanksSquareRoot(
+    const F& a, const F& quadratic_non_residue_to_trace, F* ret) {
+  // https://eprint.iacr.org/2012/685.pdf (page 12, algorithm 5)
+  // Actually this is just normal Tonelli-Shanks; since |F::GENERATOR|
+  // is a quadratic non-residue, |F::ROOT_OF_UNITY = F::GENERATOR^t|
+  // is also a quadratic non-residue (since |t| is odd).
+  if (a.IsZero()) {
+    *ret = F::Zero();
+    return true;
+  }
+  // Try computing the square root (x at the end of the algorithm)
+  // Check at the end of the algorithm if x was a square root
+  // Begin Tonelli-Shanks
+  F z = quadratic_non_residue_to_trace;
+  // w = a^((t - 1) / 2)
+  F w = a.Pow(F::Config::kTraceMinusOneDivTwo);
+  // v = s
+  size_t v = size_t{F::Config::kTwoAdicity};
+  // x = aw
+  F x = w * a;
+  // b = xw
+  F b = x * w;
+
+  while (!b.IsOne()) {
+    size_t k = 0;
+
+    // Find least integer k >= 0 such that b^(2^k) = 1.
```

Comment:
```suggestion
    // Find least integer k >= 0 such that b^(2ᵏ) = 1.
```
I'm not sure if this is a good change but maybe like this?